[Extractor] Support direct serialization and conversion to CSSValue from Style value objects

weinig opened the pull request:
#### dc119b7139d872ff07841d797dbd6122ebdcfa2a
<pre>
[Extractor] Support direct serialization and conversion to CSSValue from Style value objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=293877">https://bugs.webkit.org/show_bug.cgi?id=293877</a>

Reviewed by Darin Adler.

To serialize a Style value object for getComputedStyle purposes, adds a new `Style::serializationForCSS`
function and corresponding `Style::Serializer` specializable struct interface. This
interface differs from `CSS::serializationForCSS` in that it takes a `RenderStyle`
in addition to the existing arguments.

To convert a Style value object to a CSSValue for getComputedStyle purposes, adds a
new `Style::createCSSValue` function and corresponding `Style::CSSValueCreation`
specializable struct interface. This interface differs from `CSS::createCSSValue`
in the same way as above, in that it takes a `RenderStyle` in addition to the
existing arguments. Additionally, `CSS::CSSValueCreation` has been moved to
CSSValueTypes.h, been given a `CSSValuePool` argument and made to work generally
with ranged, tuple and variant value types.

To avoid ambiguity, a new `MinimallySerializingSpaceSeparatedSize` type has been
added for use in places where coalescing is needed. It is specialized to return
a coalescing `CSSValuePair`. All other non-&quot;MinimallySerializing&quot; tuple types
convert to a `CSSValueList`.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/borders/CSSBorderRadius.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBorderRadius.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.cpp:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueCreation.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Serialization.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/primitives/StyleURL.cpp:
* Source/WebCore/style/values/primitives/StyleURL.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/295702@main">https://commits.webkit.org/295702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da609753661ba545ce99ba760b26cba71deb134e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105911 "31 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111109 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56509 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107952 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34165 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80453 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20715 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95580 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60773 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13678 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55947 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90103 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/13714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113960 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33051 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24382 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89530 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91809 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34073 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11883 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28589 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32976 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/38387 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32722 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36071 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34320 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
```diff
@@ -100,6 +111,7 @@ auto Blending<DynamicRangeLimit>::blend(const DynamicRangeLimit& from, const Dyn
     return resolve(WTFMove(function));
 }
 
+
```
Will remove this before merging.

darinadler commented on Source/WebCore/css/values/CSSValueTypes.h:
```diff
@@ -436,6 +473,134 @@ template<> struct CSSValueChildrenVisitor<WTF::URL> {
+    else if constexpr (SerializationSeparator<T> == SerializationSeparatorType::Comma)
+        return makeCommaSeparatedListCSSValue(WTFMove(list));
+    else if constexpr (SerializationSeparator<T> == SerializationSeparatorType::Slash)
+        return makeSlashSeparatedListCSSValue(WTFMove(list));
```
I would have expected a `makeListCSSValue<SerializationSeparatorType>` for `makeListCSSValue<typename>` to call. Maybe that’s not how function templates are best used?

weinig replied:
I wanted to keep `makeSpaceSeparatedListCSSValue`, `makeCommaSeparatedListCSSValue` and `makeSlashSeparatedListCSSValue` out-of-line. 

That said, I think I can use the funky force specialization syntax to make this work with this in the header file:

```c++
template Ref<CSSValue> makeListCSSValue<SerializationSeparatorType::Space>(CSSValueListBuilder&&);
template Ref<CSSValue> makeListCSSValue<SerializationSeparatorType::Comma>(CSSValueListBuilder&&);
template Ref<CSSValue> makeListCSSValue<SerializationSeparatorType::Slash>(CSSValueListBuilder&&);
```

I will try that and see what happens.

weinig replied:
Oh, heh, that's not really what you were asking. You were asking why `typename T` vs a non-type template parameter `SerializationSeparatorType` at all. Yeah, that would be more clear.

darinadler replied:
Or maybe both, with two separate sets of function templates.

weinig committed:
[Extractor] Support direct serialization and conversion to CSSValue from Style value objects
https://bugs.webkit.org/show_bug.cgi?id=293877

Reviewed by Darin Adler.

To serialize a Style value object for getComputedStyle purposes, adds a new `Style::serializationForCSS`
function and corresponding `Style::Serializer` specializable struct interface. This
interface differs from `CSS::serializationForCSS` in that it takes a `RenderStyle`
in addition to the existing arguments.

To convert a Style value object to a CSSValue for getComputedStyle purposes, adds a
new `Style::createCSSValue` function and corresponding `Style::CSSValueCreation`
specializable struct interface. This interface differs from `CSS::createCSSValue`
in the same way as above, in that it takes a `RenderStyle` in addition to the
existing arguments. Additionally, `CSS::CSSValueCreation` has been moved to
CSSValueTypes.h, been given a `CSSValuePool` argument and made to work generally
with ranged, tuple and variant value types.

To avoid ambiguity, a new `MinimallySerializingSpaceSeparatedSize` type has been
added for use in places where coalescing is needed. It is specialized to return
a coalescing `CSSValuePair`. All other non-"MinimallySerializing" tuple types
convert to a `CSSValueList`.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/borders/CSSBorderRadius.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBorderRadius.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.cpp:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.cpp:
* Source/WebCore/style/values/color/StyleDynamicRangeLimit.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueCreation.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Serialization.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/primitives/StyleURL.cpp:
* Source/WebCore/style/values/primitives/StyleURL.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/shapes/StyleShapeFunction.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/295702@main
